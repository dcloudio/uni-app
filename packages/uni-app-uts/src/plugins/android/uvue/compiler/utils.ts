import path from 'path'
import type { Node } from '@babel/types'
import { type ExpressionNode, createSimpleExpression } from '@vue/compiler-core'
import { walk } from 'estree-walker'
import { parseExpression } from '@babel/parser'
import MagicString from 'magic-string'
import {
  camelize,
  genUTSComponentPublicInstanceIdent,
  normalizePath,
} from '@dcloudio/uni-cli-shared'
import type { TemplateCompilerOptions } from './options'
import { stringifyExpression } from './transforms/transformExpression'
import type { TransformContext } from './transform'
import type { CompilerError } from './errors'

export const __DEV__ = true
export const __BROWSER__ = false
export const __COMPAT__ = false

export function isCompatEnabled(...args: any[]) {
  return false
}

export function genRenderFunctionDecl({
  className,
  genDefaultAs,
  inline = false,
}: TemplateCompilerOptions & { genDefaultAs?: string }): string {
  if (inline) {
    return `(): any | null =>`
  }
  // 调整返回值类型为 any | null, 支持 <template>some text</template>
  const thisCode = genDefaultAs
    ? `this: InstanceType<typeof ${genDefaultAs}>`
    : ''
  return `function ${className || ''}Render(${thisCode}): any | null`
}

export function rewriteObjectExpression(
  exp: ExpressionNode,
  context: TransformContext
) {
  const source = stringifyExpression(exp)
  if (source.includes('{')) {
    const s = new MagicString(source)
    const ast = parseExpression(source, {
      plugins: context.expressionPlugins,
    })
    walk(ast, {
      enter(node: Node) {
        if (node.type === 'ObjectExpression') {
          s.prependLeft(
            node.start!,
            node.properties.length > 0 ? '_uM(' : '_uM<string, any | null>('
          )
          s.prependRight(node.end!, ')')
        }
      },
    })
    return createSimpleExpression(s.toString(), false, exp.loc)
  }
}

export function onCompilerError(error: CompilerError) {}

export function parseSource(fileName: string, rootDir: string) {
  if (fileName.includes('@dcloudio')) {
    return fileName
  }
  rootDir = normalizePath(rootDir)
  if (path.isAbsolute(fileName) && fileName.startsWith(rootDir)) {
    return '@/' + normalizePath(path.relative(rootDir, fileName))
  }
  return fileName
}

export function addEasyComponentAutoImports(
  easyComponentAutoImports: Record<string, [string, string]>,
  rootDir: string,
  tagName: string,
  fileName: string
) {
  if (easyComponentAutoImports[fileName]) {
    return
  }
  // 内置easycom，如 unicloud-db
  if (fileName.includes('@dcloudio')) {
    return
  }
  rootDir = normalizePath(rootDir)
  if (path.isAbsolute(fileName) && fileName.startsWith(rootDir)) {
    fileName = '@/' + normalizePath(path.relative(rootDir, fileName))
  }
  // 加密插件easycom类型导入
  if (fileName.includes('?uts-proxy')) {
    const moduleId = path.basename(fileName.split('?uts-proxy')[0])
    fileName = `uts.sdk.modules.${camelize(moduleId)}`
  }
  const ident = genUTSComponentPublicInstanceIdent(tagName)
  easyComponentAutoImports[fileName] = [ident, ident]
}
