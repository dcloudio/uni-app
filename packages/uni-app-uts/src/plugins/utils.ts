import path from 'path'
import { init, parse } from 'es-module-lexer'
import { normalizePath, removeExt } from '@dcloudio/uni-cli-shared'
import {
  camelize,
  capitalize,
  isArray,
  isPlainObject,
  isString,
} from '@vue/shared'

export const ENTRY_FILENAME = 'index.uts'

export async function parseImports(code: string) {
  await init
  const [imports] = parse(code)
  return imports
    .map(({ s, e }) => {
      return `import "${code.slice(s, e)}"`
    })
    .join('\n')
}

export function uvueOutDir() {
  return path.join(process.env.UNI_OUTPUT_DIR, '../.uvue')
}

export function genClassName(fileName: string, prefix: string = 'Gen') {
  return (
    prefix +
    capitalize(camelize(removeExt(normalizePath(fileName).replace(/\//g, '-'))))
  )
}

export function isVue(filename: string) {
  return filename.endsWith('.vue') || filename.endsWith('.uvue')
}

export function stringifyMap(obj: unknown) {
  return serialize(obj, true)
}

function serialize(obj: unknown, ts: boolean = false): string {
  if (isString(obj)) {
    return `"${obj}"`
  } else if (isPlainObject(obj)) {
    const entries = Object.entries(obj).map(
      ([key, value]) => `[${serialize(key, ts)},${serialize(value, ts)}]`
    )
    return `new Map${ts ? '<string, any>' : ''}([${entries.join(',')}])`
  } else if (isArray(obj)) {
    return `[${obj.map((item) => serialize(item, ts)).join(',')}]`
  } else {
    return String(obj)
  }
}
