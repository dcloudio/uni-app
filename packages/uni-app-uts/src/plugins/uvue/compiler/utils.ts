import { CompilerOptions } from './options'
import {
  ExpressionNode,
  SimpleExpressionNode,
  ConstantTypes,
} from '@vue/compiler-core'
import { isString } from '@vue/shared'

export function genRenderFunctionDecl({
  targetLanguage,
  filename,
}: CompilerOptions): string {
  return `${
    targetLanguage === 'kotlin' ? '@Suppress("UNUSED_PARAMETER") ' : ''
  }function ${filename}Render(_ctx: ${filename}): VNode | null`
}

export const objectExp = /\{.*\}/g
export function object2Map(exp: ExpressionNode | string, wrap = true) {
  const content = isString(exp) ? exp : (exp as SimpleExpressionNode).content
  const matched = content.match(objectExp)![0]
  const keyValues = matched.replace(/\{|\}/g, '').split(',')
  let mapConstructor = wrap ? 'new Map<string, any | null>([' : ''

  keyValues.forEach((keyValue: string, index: number) => {
    const colonIndex = keyValue.indexOf(':')
    const key = needAddQuotes(exp, keyValue)
      ? `'${keyValue.substring(0, colonIndex)}'`
      : keyValue.substring(0, colonIndex)
    const value = keyValue.substring(colonIndex + 1)
    if (key && value) {
      mapConstructor += `[${key},${value}]`
      if (index < keyValues.length - 1) {
        mapConstructor += ','
      }
    }
  })

  mapConstructor += wrap ? '])' : ''

  return content.replace(matched, mapConstructor)
}

function needAddQuotes(
  exp: ExpressionNode | string,
  keyValue: string
): boolean {
  return (
    !isString(exp) &&
    (exp as SimpleExpressionNode).constType === ConstantTypes.CAN_STRINGIFY &&
    !keyValue.startsWith("'") &&
    !keyValue.startsWith('"')
  )
}
