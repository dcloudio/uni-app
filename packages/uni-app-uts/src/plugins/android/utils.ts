import fs from 'fs'
import path from 'path'
import { type ImportSpecifier, init, parse } from 'es-module-lexer'
import {
  createResolveErrorMsg,
  createRollupError,
  getUTSEasyComAutoImports,
  normalizeNodeModules,
  offsetToStartAndEnd,
  parseUniExtApiNamespacesJsOnce,
} from '@dcloudio/uni-cli-shared'
import { isArray, isPlainObject, isString } from '@vue/shared'

import { type Import, createUnimport } from 'unimport'

import type { /*SourceMapInput, */ PluginContext } from 'rollup'
import type { Position, SourceLocation } from '@vue/compiler-core'

import { createCompilerError } from './uvue/compiler/errors'

export const UVUE_CLASS_NAME_PREFIX = 'Gen'

export const DEFAULT_APPID = '__UNI__uniappx'

export const ENTRY_FILENAME = () =>
  process.env.UNI_APP_X_TSC === 'true' ? 'main.uts.ts' : 'main.uts'

export function wrapResolve(
  resolve: PluginContext['resolve']
): PluginContext['resolve'] {
  return async (source, importer, options) => {
    try {
      return await resolve(source, importer, options)
    } catch (e) {
      // import "@/pages/logo.png" 可能会报 Cannot find module 错误
    }
    return null
  }
}

export function createTryResolve(
  importer: string,
  resolve: PluginContext['resolve'],
  offsetStart?: Position,
  origCode: string = ''
) {
  return async (
    source: string,
    code: string,
    { ss, se }: ImportSpecifier
  ): Promise<boolean | void> => {
    const resolved = await wrapResolve(resolve)(source, importer)
    if (!resolved) {
      const { start, end } = offsetToStartAndEnd(code, ss, se)
      if (offsetStart) {
        if (start.line === 1) {
          start.column = start.column + offsetStart.column
          if (end.line === 1) {
            end.column = end.column + offsetStart.column
          }
        }
        const offsetLine = offsetStart.line - 1
        start.line = start.line + offsetLine
        end.line = end.line + offsetLine
      }
      throw createResolveError(
        origCode || code,
        createResolveErrorMsg(source, importer),
        start,
        end
      )
    }
  }
}

export async function parseImports(
  code: string,
  tryResolve?: ReturnType<typeof createTryResolve>
) {
  await init
  let res: ReturnType<typeof parse> = [[], [], false, false]
  try {
    res = parse(code)
  } catch (err: any) {
    const message = err.message
    if (message) {
      const matches = message.match(/@:(\d+):(\d+)/)
      if (matches) {
        throw createRollupError(
          '',
          '',
          createCompilerError(
            0,
            {
              start: {
                offset: 0,
                line: parseInt(matches[1]),
                column: parseInt(matches[2]),
              },
            } as SourceLocation,
            { 0: `Parse error` },
            ''
          ),
          code
        )
      }
    }
    throw err
  }
  const imports = res[0]
  const importsCode: string[] = []
  for (const specifier of imports) {
    const source = code.slice(specifier.s, specifier.e)
    if (tryResolve) {
      const res = await tryResolve(source, code, specifier)
      if (res === false) {
        // 忽略该import
        continue
      }
    }
    importsCode.push(`import "${source}"`)
  }

  return importsCode.concat(parseUniExtApiImports(code)).join('\n')
}

export function createResolveError(
  code: string,
  msg: string,
  start: Position,
  end: Position
) {
  return createRollupError(
    '',
    '',
    createCompilerError(
      0,
      {
        start,
        end,
      } as SourceLocation,
      { 0: msg },
      ''
    ),
    code
  )
}

// @ts-expect-error 暂时不用
function genImportsCode(code: string, imports: readonly ImportSpecifier[]) {
  const chars = code.split('')
  const keepChars: number[] = []
  imports.forEach(({ ss, se }) => {
    for (let i = ss; i <= se; i++) {
      keepChars.push(i)
    }
  })
  for (let i = 0; i < chars.length; i++) {
    if (!keepChars.includes(i)) {
      const char = chars[i]
      if (char !== '\r' && char !== '\n') {
        chars[i] = ' '
      }
    }
  }
  return chars.join('')
}

function parseUniExtApiImports(code: string): string[] {
  if (!process.env.UNI_UTS_PLATFORM) {
    return []
  }
  const extApis = parseUniExtApiNamespacesJsOnce(
    process.env.UNI_UTS_PLATFORM,
    process.env.UNI_UTS_TARGET_LANGUAGE
  )
  const pattern = /uni\.(\w+)/g
  const apis = new Set<string>()
  let match
  while ((match = pattern.exec(code)) !== null) {
    apis.add(match[1])
  }
  const imports: string[] = []
  apis.forEach((api) => {
    const extApi = extApis[api]
    if (extApi) {
      imports.push(`import "${extApi[0]}"`)
    }
  })
  return imports
}

export function kotlinOutDir() {
  return path.join(process.env.UNI_OUTPUT_DIR, '../.kotlin')
}

export function isVue(filename: string) {
  return filename.endsWith('.vue') || filename.endsWith('.uvue')
}

export function stringifyMap(obj: unknown, ts = false) {
  return serialize(obj, ts)
}

function serialize(obj: unknown, ts = false): string {
  if (isString(obj)) {
    return `"${obj}"`
  } else if (isPlainObject(obj)) {
    const entries = Object.entries(obj).map(
      ([key, value]) => `[${serialize(key, ts)},${serialize(value, ts)}]`
    )
    if (entries.length) {
      return `utsMapOf([${entries.join(',')}])`
    }
    if (ts) {
      return `utsMapOf<string, any | null>()`
    }
    return `utsMapOf()`
  } else if (isArray(obj)) {
    return `[${obj.map((item) => serialize(item, ts)).join(',')}]`
  } else {
    return String(obj)
  }
}

export function parseUTSRelativeFilename(filename: string, root?: string) {
  if (!path.isAbsolute(filename)) {
    return filename
  }
  return normalizeNodeModules(
    path.relative(root ?? process.env.UNI_INPUT_DIR, filename)
  )
}

export function parseUTSImportFilename(filename: string) {
  if (!path.isAbsolute(filename)) {
    return filename
  }
  return (
    '@/' +
    normalizeNodeModules(path.relative(process.env.UNI_INPUT_DIR, filename))
  )
}

type UniCloudSpace = {
  provider: string
  spaceName: string
  spaceId: string
  clientSecret?: string
  endpoint?: string
  workspaceFolder?: string
}

let uniCloudSpaceList: Array<UniCloudSpace>
export function getUniCloudSpaceList(): Array<UniCloudSpace> {
  if (uniCloudSpaceList) {
    return uniCloudSpaceList
  }
  if (!process.env.UNI_CLOUD_SPACES) {
    uniCloudSpaceList = []
    return uniCloudSpaceList
  }
  try {
    const spaces = JSON.parse(process.env.UNI_CLOUD_SPACES)
    if (!Array.isArray(spaces)) {
      uniCloudSpaceList = []
      return uniCloudSpaceList
    }
    uniCloudSpaceList = spaces.map((space) => {
      if (space.provider === 'tcb') {
        space.provider = 'tencent'
      }
      if (!space.provider && space.clientSecret) {
        space.provider = 'aliyun'
      }
      switch (space.provider) {
        case 'aliyun':
        case 'dcloud':
          return {
            provider: space.provider || 'aliyun',
            spaceName: space.name,
            spaceId: space.id,
            clientSecret: space.clientSecret,
            endpoint: space.apiEndpoint,
            workspaceFolder: space.workspaceFolder,
          }
        case 'alipay': {
          return {
            provider: space.provider,
            spaceName: space.name,
            spaceId: space.id,
            spaceAppId: space.spaceAppId,
            accessKey: space.accessKey,
            secretKey: space.secretKey,
            workspaceFolder: space.workspaceFolder,
          }
        }
        case 'tencent':
        default: {
          return {
            provider: space.provider,
            spaceName: space.name,
            spaceId: space.id,
            workspaceFolder: space.workspaceFolder,
          }
        }
      }
    })
  } catch (e) {
    console.error(e)
  }
  uniCloudSpaceList = uniCloudSpaceList || []
  if (uniCloudSpaceList.length > 1) {
    console.warn('Multi uniCloud space is not supported yet.')
  }
  return uniCloudSpaceList
}

// TODO copy from uni-uts-v1 compiler, should be refactor
type UniCloudObjectInfo = {
  name: string
  methodList: string[]
}

export function getUniCloudObjectInfo(
  uniCloudSpaceList: Array<UniCloudSpace>
): Array<UniCloudObjectInfo> {
  let uniCloudWorkspaceFolder = process.env.UNI_INPUT_DIR.endsWith('src')
    ? path.resolve(process.env.UNI_INPUT_DIR, '..')
    : process.env.UNI_INPUT_DIR
  let serviceProvider = 'aliyun'
  if (uniCloudSpaceList && uniCloudSpaceList.length > 0) {
    const space = uniCloudSpaceList[0]
    if (space.workspaceFolder) {
      uniCloudWorkspaceFolder = space.workspaceFolder
    }
    serviceProvider = space.provider === 'tencent' ? 'tcb' : space.provider
  } else {
    serviceProvider =
      ['aliyun', 'tcb', 'alipay', 'dcloud'].find((item) =>
        fs.existsSync(path.resolve(uniCloudWorkspaceFolder, 'uniCloud-' + item))
      ) || 'aliyun'
  }
  try {
    const { getWorkspaceObjectInfo } = require('../../../lib/unicloud-utils')
    return getWorkspaceObjectInfo(uniCloudWorkspaceFolder, serviceProvider)
  } catch (e) {
    console.error((e as Error).message)
    process.exit(1)
  }
}

export async function transformAutoImport(
  code: string,
  id: string,
  ignore: string[] = []
) {
  const { matchedImports } = await detectAutoImports(code, id, ignore)
  if (matchedImports.length) {
    return {
      code: code + '\n' + genAutoImportsCode(matchedImports),
    }
  }
  return { code }
}

export function genAutoImportsCode(imports: Import[]) {
  const codes: string[] = []
  imports.forEach(({ name, as, from }) => {
    if (as && name !== as) {
      codes.push(
        `import { ${name} as ${as} } from "${parseUTSImportFilename(from)}"`
      )
    } else {
      codes.push(`import { ${name}  } from "${parseUTSImportFilename(from)}"`)
    }
  })
  return codes.join('\n')
}

let detectImports: DetectImports

export function transformUniCloudMixinDataCom(code: string) {
  // 将 uniCloud.mixinDatacom 替换为 uniCloudMixinDatacom
  // 然后 autoImport 会自动导入 uniCloudMixinDatacom
  if (code.includes('uniCloud.mixinDatacom')) {
    return code.replace(/uniCloud\.mixinDatacom/g, 'uniCloudMixinDatacom')
  }
  return code
}

export function detectAutoImports(
  code: string,
  id: string,
  ignore: string[] = []
) {
  // 目前硬编码
  if (id.includes('index.module.uts')) {
    return { matchedImports: [] }
  }
  if (!detectImports) {
    detectImports = initAutoImport().detectImports
  } else {
    const autoImports = getUTSEasyComAutoImports()
    const sources = Object.keys(autoImports)
    if (detectImports.key !== sources.sort().join(',')) {
      detectImports = initAutoImport().detectImports
    }
  }

  return detectImports(code, id, ignore)
}

type DetectImports = {
  key: string
  (code: string, id: string, ignore: string[]): Promise<{
    matchedImports: Import[]
  }>
}

function initAutoImport(): {
  detectImports: DetectImports
} {
  const autoImports = getUTSEasyComAutoImports()
  const sources = Object.keys(autoImports)
  if (!sources.length) {
    const detectImports = async (
      _code: string,
      _id: string,
      _ignore: string[] = []
    ) => {
      return { matchedImports: [] }
    }
    detectImports.key = 'default'
    return {
      detectImports,
    }
  }

  const imports: Import[] = []
  sources.forEach((source) => {
    autoImports[source].forEach(([name, as]) => {
      imports.push({
        name,
        as,
        from: source,
      })
    })
  })
  const { detectImports: uniDetectImports } = createUnimport({
    imports,
  })
  const detectImports = async function (
    code: string,
    id: string,
    ignore: string[] = []
  ) {
    // const start = Date.now()
    const result = await uniDetectImports(code)
    // console.log('detectImports[' + id + ']耗时:' + (Date.now() - start))
    return {
      matchedImports: result.matchedImports.filter((item) => {
        if (item.as && item.name !== item.as) {
          return !ignore.includes(item.as)
        }
        return !ignore.includes(item.name)
      }),
    }
  }
  detectImports.key = sources.sort().join(',')
  return {
    detectImports,
  }
}
