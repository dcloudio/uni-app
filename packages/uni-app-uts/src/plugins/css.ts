import type { Plugin, ResolvedConfig } from 'vite'
import path from 'path'
import colors from 'picocolors'

import {
  commonjsProxyRE,
  cssLangRE,
  cssPlugin,
  cssPostPlugin,
  formatAtFilename,
  generateCodeFrame,
  insertBeforePlugin,
  normalizePath,
  parseVueRequest,
  resolveMainPathOnce,
} from '@dcloudio/uni-cli-shared'
import { parse } from '@dcloudio/uni-nvue-styler'

import { genClassName, isVue } from './utils'

export function uniAppCssPlugin(): Plugin {
  const mainUTS = resolveMainPathOnce(process.env.UNI_INPUT_DIR)
  let resolvedConfig: ResolvedConfig
  const name = 'uni:app-uvue-css'
  return {
    name,
    apply: 'build',
    configResolved(config) {
      resolvedConfig = config
      const uvueCssPostPlugin = cssPostPlugin(config, {
        isJsCode: true,
        platform: process.env.UNI_PLATFORM,
        chunkCssFilename(id: string) {
          if (id === mainUTS) {
            return 'App.vue.style.uts'
          }
          const { filename } = parseVueRequest(id)
          if (isVue(filename)) {
            return normalizePath(
              path.relative(process.env.UNI_INPUT_DIR, filename) + '.style.uts'
            )
          }
        },
        async chunkCssCode(filename, cssCode) {
          const { code, messages } = await parse(cssCode, {
            filename,
            logLevel: 'ERROR',
            map: true,
            ts: true,
            type: 'uvue',
          })
          messages.forEach((message) => {
            if (message.type === 'error') {
              let msg = `[plugin:uni:app-uvue-css] ${message.text}`
              if (message.line && message.column) {
                msg += `\n${generateCodeFrame(cssCode, {
                  line: message.line,
                  column: message.column,
                }).replace(/\t/g, ' ')}`
              }
              msg += `\n${formatAtFilename(filename)}`
              resolvedConfig.logger.error(colors.red(msg))
            }
          })
          return `export const ${genClassName(
            filename.replace('.style.uts', '')
          )}Styles = [${code}]`
        },
      })
      // 增加 css plugins
      insertBeforePlugin(cssPlugin(config), name, config)
      const plugins = config.plugins as Plugin[]
      const index = plugins.findIndex((p) => p.name === 'uni:app-uvue')
      plugins.splice(index, 0, uvueCssPostPlugin)
    },
    async transform(source, filename) {
      if (!cssLangRE.test(filename) || commonjsProxyRE.test(filename)) {
        return
      }
      // 仅做校验使用
      const { messages } = await parse(source, {
        filename,
        logLevel: 'WARNING',
        map: true,
        ts: true,
        noCode: true,
        type: 'uvue',
      })
      messages.forEach((message) => {
        if (message.type === 'warning') {
          let msg = `[plugin:uni:app-uvue-css] ${message.text}`
          if (message.line && message.column) {
            msg += `\n${generateCodeFrame(source, {
              line: message.line,
              column: message.column,
            }).replace(/\t/g, ' ')}`
          }
          msg += `\n${formatAtFilename(filename)}`
          resolvedConfig.logger.warn(colors.yellow(msg))
        }
      })
      return { code: source }
    },
  }
}
