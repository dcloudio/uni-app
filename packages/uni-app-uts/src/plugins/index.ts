import path from 'path'
import fs from 'fs-extra'
import {
  UniViteCopyPluginOptions,
  UniVitePlugin,
  emptyDir,
  initI18nOptions,
  normalizeNodeModules,
  normalizePath,
  parseManifestJsonOnce,
  parseVueRequest,
  resolveMainPathOnce,
  resolveUTSCompiler,
  utsPlugins,
} from '@dcloudio/uni-cli-shared'
import { compileI18nJsonStr } from '@dcloudio/uni-i18n'
import type { Plugin } from 'vite'
import { parseImports, uvueOutDir } from './utils'

const REMOVED_PLUGINS = [
  'vite:build-metadata',
  'vite:modulepreload-polyfill',
  'vite:css',
  'vite:esbuild',
  'vite:json',
  'vite:wasm-helper',
  'vite:worker',
  'vite:asset',
  'vite:wasm-fallback',
  'vite:define',
  'vite:css-post',
  'vite:build-html',
  'vite:html-inline-proxy',
  'vite:worker-import-meta-url',
  'vite:asset-import-meta-url',
  'vite:force-systemjs-wrap-complete',
  'vite:watch-package-data',
  'commonjs',
  'vite:data-uri',
  'vite:dynamic-import-vars',
  'vite:import-glob',
  'vite:build-import-analysis',
  'vite:esbuild-transpile',
  'vite:terser',
  'vite:reporter',
]

export function uniAppUTSPlugin(): UniVitePlugin {
  const inputDir = process.env.UNI_INPUT_DIR
  const outputDir = process.env.UNI_OUTPUT_DIR
  const mainUTS = resolveMainPathOnce(inputDir)
  const tempOutputDir = uvueOutDir()
  const manifestJson = parseManifestJsonOnce(inputDir)

  // 开始编译时，清空输出目录
  function emptyOutDir() {
    if (fs.existsSync(outputDir)) {
      emptyDir(outputDir)
    }
  }
  emptyOutDir()
  return {
    name: 'uni:app-uts',
    apply: 'build',
    uni: createUniOptions(),
    config() {
      return {
        base: '/', // 强制 base
        build: {
          outDir: tempOutputDir,
          lib: {
            // 必须使用 lib 模式
            fileName: 'output',
            entry: resolveMainPathOnce(inputDir),
            formats: ['cjs'],
          },
          rollupOptions: {
            external(source) {
              if (['vue', 'vuex', 'pinia'].includes(source)) {
                return true
              }
              // 相对目录
              if (source.startsWith('@/') || source.startsWith('.')) {
                return false
              }
              if (path.isAbsolute(source)) {
                return false
              }
              // android 系统库，三方库
              if (source.includes('.')) {
                return true
              }
              return false
            },
          },
        },
      }
    },
    configResolved(config) {
      const plugins = config.plugins as Plugin[]
      const len = plugins.length
      for (let i = len - 1; i >= 0; i--) {
        const plugin = plugins[i]
        if (REMOVED_PLUGINS.includes(plugin.name)) {
          plugins.splice(i, 1)
        }
      }
    },
    async transform(code, id) {
      const { filename } = parseVueRequest(id)
      if (!filename.endsWith('.uts')) {
        return
      }
      // 仅处理 uts 文件
      // 忽略 uni-app-uts/lib/automator/index.uts
      if (!filename.includes('uni-app-uts')) {
        const isMainUTS = normalizePath(id) === mainUTS
        const fileName = path.relative(inputDir, id)
        this.emitFile({
          type: 'asset',
          fileName: normalizeFilename(fileName, isMainUTS),
          source: normalizeCode(code, isMainUTS),
        })
      }
      code = await parseImports(code)
      return code
    },
    async writeBundle() {
      const res = await resolveUTSCompiler().compileApp(
        path.join(tempOutputDir, 'index.uts'),
        {
          inputDir: tempOutputDir,
          outputDir: outputDir,
          package: 'uni.' + (manifestJson.appid || '').replace(/_/g, ''),
          sourceMap: true,
          uni_modules: [...utsPlugins],
        }
      )
      if (res) {
        const files: string[] = []
        if (process.env.UNI_APP_UTS_CHANGED_FILES) {
          try {
            files.push(...JSON.parse(process.env.UNI_APP_UTS_CHANGED_FILES))
          } catch (e) {}
        }
        if (res.changed && res.changed.length) {
          files.push(...res.changed)
        }
        process.env.UNI_APP_UTS_CHANGED_FILES = JSON.stringify([
          ...new Set(files),
        ])
      }
    },
  }
}

function normalizeFilename(filename: string, isMain = false) {
  if (isMain) {
    return 'index.uts'
  }
  return normalizeNodeModules(filename)
}

function normalizeCode(code: string, isMain = false) {
  if (!isMain) {
    return code
  }
  const automatorCode = process.env.UNI_AUTOMATOR_WS_ENDPOINT
    ? 'initAutomator();'
    : ''
  return `
${code}  
export function main(app: IApp) {
    defineAppConfig();
    definePageRoutes();
    ${automatorCode}
    (createApp()['app'] as VueApp).mount(app);
}
`
}

function createUniOptions(): UniVitePlugin['uni'] {
  return {
    copyOptions() {
      const platform = process.env.UNI_PLATFORM
      const inputDir = process.env.UNI_INPUT_DIR
      const outputDir = process.env.UNI_OUTPUT_DIR
      const targets: UniViteCopyPluginOptions['targets'] = []
      // 自动化测试时，不启用隐私政策
      if (!process.env.UNI_AUTOMATOR_WS_ENDPOINT) {
        targets.push({
          src: 'androidPrivacy.json',
          dest: outputDir,
          transform(source) {
            const options = initI18nOptions(platform, inputDir, false, true)
            if (!options) {
              return
            }
            return compileI18nJsonStr(source.toString(), options)
          },
        })
        const debugFilename = '__nvue_debug__'
        if (fs.existsSync(path.resolve(inputDir, debugFilename))) {
          targets.push({
            src: debugFilename,
            dest: outputDir,
          })
        }
      }
      return {
        assets: ['hybrid/html/**/*', 'uni_modules/*/hybrid/html/**/*'],
        targets,
      }
    },
  }
}
