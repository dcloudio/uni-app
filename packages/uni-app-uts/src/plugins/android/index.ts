import {
  isNormalCompileTarget,
  parseUniExtApiNamespacesOnce,
  uniDecryptUniModulesPlugin,
  uniEncryptUniModulesPlugin,
  uniUTSAppUniModulesPlugin,
  uniUniModulesExtApiPlugin,
  uniViteSfcSrcImportPlugin,
} from '@dcloudio/uni-cli-shared'
import { uniPrePlugin } from './pre'
import { uniAppPlugin } from './plugin'
import { uniAppCssPlugin, uniAppCssPrePlugin } from './css'
import { uniAppMainPlugin } from './mainUTS'
import { uniAppManifestPlugin } from './manifestJson'
import { uniAppPagesPlugin } from './pagesJson'
import { uniAppUVuePlugin } from './uvue'
import { uniCloudPlugin } from './unicloud'

export function init() {
  return [
    uniAppCssPrePlugin(),
    ...(isNormalCompileTarget()
      ? [
          require('@dcloudio/uni-console/lib/uni.plugin.js')(),
          uniDecryptUniModulesPlugin(),
        ]
      : []),
    uniPrePlugin(),
    ...(isNormalCompileTarget()
      ? [
          uniUTSAppUniModulesPlugin({
            x: true,
            isSingleThread: process.env.UNI_APP_X_SINGLE_THREAD !== 'false',
            extApis: parseUniExtApiNamespacesOnce(
              process.env.UNI_UTS_PLATFORM,
              process.env.UNI_UTS_TARGET_LANGUAGE
            ),
          }),
        ]
      : []),
    uniAppPlugin(),
    ...(process.env.UNI_COMPILE_TARGET === 'ext-api'
      ? [uniUniModulesExtApiPlugin()]
      : process.env.UNI_COMPILE_TARGET === 'uni_modules'
      ? [uniEncryptUniModulesPlugin()]
      : [
          // 需要放到 uniAppPlugin 之后(TSC模式无需)，让 uniAppPlugin 先 emit 出真实的 main.uts，然后 MainPlugin 再返回仅包含 import 的 js code
          uniAppMainPlugin(),
          uniAppManifestPlugin(),
          uniAppPagesPlugin(),
        ]),

    uniAppCssPlugin(),
    // 解决所有的src引入
    uniViteSfcSrcImportPlugin({ onlyVue: false }),
    uniAppUVuePlugin(),
    uniCloudPlugin(),
  ]
}
