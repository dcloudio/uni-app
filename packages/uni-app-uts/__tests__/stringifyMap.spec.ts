import { stringifyMap } from '../src/plugins/utils'

describe('stringifyMap', () => {
  test(`basic`, () => {
    expect(
      stringifyMap({
        color: '#7A7E83',
        selectedColor: '#3cc51f',
        borderStyle: 'black',
        backgroundColor: '#ffffff',
        list: [
          {
            pagePath: 'pages/component/index',
            iconPath: 'static/image/icon_component.png',
            selectedIconPath: 'static/image/icon_component_HL.png',
            text: '组件',
          },
          {
            pagePath: 'pages/API/index',
            iconPath: 'static/image/icon_API.png',
            selectedIconPath: 'static/image/icon_API_HL.png',
            text: '接口',
          },
        ],
      })
    ).toMatchSnapshot()
  })
})
