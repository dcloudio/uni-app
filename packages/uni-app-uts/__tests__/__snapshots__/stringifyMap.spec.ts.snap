// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringifyMap basic 1`] = `"new Map<string, any>([["color","#7A7E83"],["selectedColor","#3cc51f"],["borderStyle","black"],["backgroundColor","#ffffff"],["list",[new Map<string, any>([["pagePath","pages/component/index"],["iconPath","static/image/icon_component.png"],["selectedIconPath","static/image/icon_component_HL.png"],["text","组件"]]),new Map<string, any>([["pagePath","pages/API/index"],["iconPath","static/image/icon_API.png"],["selectedIconPath","static/image/icon_API_HL.png"],["text","接口"]])]]])"`;
